### PR TITLE
[Feature:UI/UX] Togglable dark theme

### LINKED ISSUE #1873: Feature Request: Dark Theme
Like many other dark-dwelling grad students, I prefer dark themes when looking at code. The light color theme of Submitty strains my eyes, and I would appreciate a dark theme.

I have tried adjusting the colors in my browser, but that overrides Submitty's syntax coloring.

Thanks!

### PR DESCRIPTION
**Pull request by elihschiff**
### Please check if the PR fulfills these requirements:

* [X] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [X] Tests for the changes have been added/updated (if possible)

### What is the current behavior?

Submitty only has a light theme
closes #1873

### What is the new behavior?

I added a button on the sidebar to toggle between light theme and dark theme.
image
image
Here are some example pages:
image
image
image
image
image
image
image
image
image

### Other information?

For testers:
I tried not to mess with the light theme colors at all but there is a chance I messed something up so make sure everything still looks good in the normal light theme
For the dark theme, I tried to make everything able to be read but I am sure I missed something so feel free to let me know if you find anything that is hard to see or is the wrong color on dark theme.

Also if somebody has other colors they think would look better in the dark theme feel free to leave feedback.

### CODE REVIEW

File: site/public/js/server.js [hunk 1]
```diff
@@ -1576,6 +1576,28 @@ function checkSidebarCollapse() {
...
+  if(mode==='normal'){
+    if((!localStorage.getItem("theme") && document.documentElement.getAttribute("data-theme") != "dark") || localStorage.getItem("theme") == "light"){
+        localStorage.setItem("theme", "dark");
+        document.documentElement.setAttribute("data-theme", "dark");
+    }else{
+      localStorage.setItem("theme", "light");
+      document.documentElement.setAttribute("data-theme", "light");
+    }
+  }else if(mode === 'black'){
+    if(!localStorage.getItem("black_mode") || localStorage.getItem("black_mode") != "black"){
```
Comment (MasterOdin) [inline site/public/js/server.js]: Can just do `localStorage.getItem('black_mode') !== 'black'` for test, as this covers both falsy values and incorrect string values.

File: site/app/templates/GlobalHeader.twig [hunk 2]
```diff
@@ -61,6 +61,32 @@
         }
 
         var onAjaxInit;
+
+
+        //determines if the user has a set theme
+        function detectColorScheme(){
+            var theme="light";    //default to light
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: Use [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)

File: site/app/templates/GlobalHeader.twig [hunk 3]
```diff
@@ -61,6 +61,32 @@
...
+            //local storage is used to override OS theme settings
+            if(localStorage.getItem("theme")){
+                if(localStorage.getItem("theme") == "dark"){
+                    var theme = "dark";
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: get rid of the `var` here, it's already been declared above.

File: site/app/templates/GlobalHeader.twig [hunk 4]
```diff
@@ -61,6 +61,32 @@
...
+                }
+            } else if(!window.matchMedia) {
+                //matchMedia method not supported
+                return false;
+            } else if(window.matchMedia("(prefers-color-scheme: dark)").matches) {
+                //OS theme setting detected as dark
+                var theme = "dark";
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: get rid of the `var` here, it's already been declared above.

File: site/app/templates/GlobalHeader.twig [hunk 5]
```diff
@@ -61,6 +61,32 @@
...
 
         var onAjaxInit;
+
+
+        //determines if the user has a set theme
+        function detectColorScheme(){
+            var theme="light";    //default to light
+            //local storage is used to override OS theme settings
+            if(localStorage.getItem("theme")){
+                if(localStorage.getItem("theme") == "dark"){
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: use `===` when testing for equality

File: site/app/templates/GlobalHeader.twig [hunk 6]
```diff
@@ -61,6 +61,32 @@
...
+                }
+            } else if(!window.matchMedia) {
+                //matchMedia method not supported
+                return false;
+            } else if(window.matchMedia("(prefers-color-scheme: dark)").matches) {
+                //OS theme setting detected as dark
+                var theme = "dark";
+            }
+            //dark theme preferred, set document with a `data-theme` attribute
+            if (theme=="dark") {
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: use `===` when testing for equality

File: site/app/templates/GlobalHeader.twig [hunk 7]
```diff
@@ -61,6 +61,32 @@
...
+                 document.documentElement.setAttribute("data-theme", "dark");
+                 if(localStorage.getItem("black_mode") == "black"){
```
Comment (MasterOdin) [inline site/app/templates/GlobalHeader.twig]: use `===` when testing for equality

File: site/app/templates/HomePage.twig [hunk 8]
```diff
@@ -56,6 +56,28 @@
                     </a>
                 </li>
             {% endif %}
+            <li>
+                <br>
+                <button title="Enable Dark Mode" class="btn btn-block dark_mode_toggle key_to_click" onclick="toggleTheme();" style="background-color: var(--dark-mode-black); color: var(--text-white)">
```
Comment (MasterOdin) [inline site/app/templates/HomePage.twig]: can we move the style attributes here into the `dark_mode_toggle` class (and same with the other buttons and their style attributes.

File: site/public/js/server.js [hunk 9]
```diff
@@ -1576,6 +1576,28 @@ function checkSidebarCollapse() {
     }
 }
 
+//Changes the theme from light to dark mode or the reverse
+//if mode='black' it will toggle the black mode instead of the normal mode
+function toggleTheme(mode='normal'){
+  if(mode==='normal'){
+    if((!localStorage.getItem("theme") && document.documentElement.getAttribute("data-theme") != "dark") || localStorage.getItem("theme") == "light"){
```
Comment (MasterOdin) [inline site/public/js/server.js]: use `!==` and `===` for testing for equality

**Comment by matgar1997** [comment]
Screen Shot 2020-03-19 at 10 31 38 PM

Small issue, in dark mode if a white button is selected then you go back a page the button becomes hard to see

**Comment by elihschiff** [comment]
The only vpat issue this would cause is contrast issues. The way we have been currently checking that is using wave however their api costs money so we cannot really set up a test for it.

The one idea I did have was maybe we can sideload the extension in selenium and run it ourselves using selenium but that is something I would need to talk more about at the meeting and I am not even sure it is doable even if we wanted to. That would need to be its own future pr because it would need a lot of work to get it working correctly.

I missed that black when you click a button, I can fix that

**Comment by matgar1997** [comment]
A couple of other contrast issues, some ofthe buttons become really hard to see. The second is the settings page within the NavigationBar
Screen Shot 2020-03-19 at 11 42 43 PM
Screen Shot 2020-03-19 at 11 41 00 PM

**Comment by elihschiff** [comment]
The colors for the rows in that settings box as well as the buttons should now be fixed. Good catch

**Comment by bmcutler** [comment]
@elihschiff , did you move this from the sidebar to an option under "about me" yet?
Also, I'm going to delay merging this until you make a documentation PR (or bribe @matgar1997 to do it for you)

**Comment by MasterOdin** [comment]
@bmcutler he did
Screenshot 2020-03-26 14 18 34

**Comment by bmcutler** [comment]
Two big buttons is sort of huge.  Can we make it a checkbox or radio button?

**Comment by bmcutler** [comment]
We might eventually have a small number of other options users can set in this space.

For example if we were to allow users to specify their timezone (defaulting to what we read from their system clock) it could be displayed & changed  here.

**Comment by elihschiff** [comment]
image
image
@bmcutler how about this?

**Comment by bmcutler** [comment]
what's dark vs black?  I thought it was just a single boolean

**Comment by elihschiff** [comment]
If you have a certain type of screen, having something that is 100% black will actually save battery life. However if your screen is not of that type it often looks worse to have a 100% black background. Therefore most places give you the option. The only difference is the background color.

Dark mode: 
image
Black mode: 
image

**Comment by bmcutler** [comment]
ok, thanks for the explanation :)   I approve.
write the docs and then I'll merge...

**Pull request merged**